[cli] Add dry run for client commands

stefan-mysten opened the pull request:
## Description 

This PR adds the `dry-run` flag for the following `sui client` commands:
- call
- merge-coin
- split-coin
- pay
- pay-sui
- pay-all-sui
- transfer
- transfer-sui
- publish
- upgrade
- ptb

Note that the `gas-budget` is still required to be passed despite passing `dry-run`. That is intentional and it will be changed to become optional in a subsequent PR that adds automatic gas budget estimation.

## Test Plan 

Existing tests

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
 Added the `dry-run` flag for the following `sui client` commands:
- call
- merge-coin
- split-coin
- pay
- pay-sui
- pay-all-sui
- transfer
- transfer-sui
- publish
- upgrade
- ptb

amnn commented on crates/sui/src/client_commands.rs:
For a later clean-up, you can pull out the common flags across all these commands (`gas`, `gas-budget`, `dry-run`, `gas-price`, `serialize-unsigned-transaction`, `serialize-signed-transaction`) into their own type and flatten them into each client command variant, to avoid all this duplication.

stefan-mysten replied:
That sounds good!

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -987,19 +1032,40 @@ impl SuiClientCommands {
+                    )
+                    .await?;
+                if dry_run {
+                    return helper_execute_dry_run(
```
nit: we can omit `helper` and `execute` from the name here -- `helper` I don't think is adding much, and `execute` is a little confusing because that's how we refer to running a transaction for its effects.

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -1173,18 +1266,57 @@ impl SuiClientCommands {
                 .await?;
+                let gas_owner = context.try_get_object_owner(&gas).await?;
+                let signer = gas_owner.unwrap_or(context.active_address()?);
+                let gas_price = gas_price.unwrap_or(context.get_reference_gas_price().await?);
```
To avoid the API call if we don't need it:

```suggestion
                let gas_price = if let Some(gas_price) {
                    gas_price
                } else 
                    context.get_reference_gas_price().await?
                };
```

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -1173,18 +1266,57 @@ impl SuiClientCommands {
+                    )
+                    .await;
+                }
+                // This is what it's being doen in the original move_call code,
+                // and it is passed to the select_gas function.
```
nit: typos in comment

amnn replied:
Can we generalise the implementation of `tx_data` so that it extracts the input objects from `tx_kind` itself? It seems like that should be doable.

stefan-mysten replied:
Not sure I follow here. In this one place (`Call` command), the `input_coins` are taken out of the tx_kind, but in other commands, `input_coins` are passed as arguments.

stefan-mysten replied:
Oh, now I see, I think we can do that.

stefan-mysten replied:
I would do that though in a separate PR. I am not confident about one part of it.

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -1272,9 +1456,43 @@ impl SuiClientCommands {
+
+                if let Some(gas) = gas {
+                    if input_coins.contains(&gas) {
+                        return Err(anyhow!("Gas coin is in input coins of Pay transaction, use PaySui transaction instead!"));
```
nit: use `bail!(...)` instead of `return Err(anyhow!(...))`

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -1459,31 +1729,50 @@ impl SuiClientCommands {
-                    _ => {
-                        return Err(anyhow!("Exactly one of `count` and `amounts` must be present for split-coin command."));
+                if amounts.is_some() && count.is_some() {
+                    return Err(anyhow!("Exactly one of `count` and `amounts` must be present for split-coin command."));
+                }
+                if let Some(count) = count {
+                    if count == 0 {
+                        return Err(anyhow!("Coin split count must be greater than 0"));
                     }
-                };
+                }
```
nit (optional): Could be a single `match` command (also what happens if they are both `None`?

```rust
match (count, amounts) {
  (Some(0), _) => bail!("Coin split count must be greater than 0"),
  (Some(_), Some(_)) | (None, None) => bail!("Exactly one of `count` and `amounts`...");
  (Some(_), None) | (None, Some(_)) => { /* nop */ },
``

stefan-mysten replied:
Oops, forgot that case.

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -97,6 +97,79 @@ impl TransactionBuilder {
         }
     }
 
+    /// Construct the transaction data for a dry run (uses min gas budget)
```
I'm not sure what the "uses min gas budget" part means here, because this function is being passed a gas budget.

stefan-mysten replied:
Ah, forgot to remove this oudated comment.

amnn replied:
This comment still seems to be around, but was marked as resolved -- are there changes that need to be pushed still?

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -97,6 +97,79 @@ impl TransactionBuilder {
+        )
+    }
+
+    /// Construct the transaction data from a transaction kind and a few needed
+    /// Use the `input_coins` for pay_sui
```
This comment doesn't make a whole lot of sense to me.

stefan-mysten replied:
Oops, something went wrong here / forgot to fix it. Thanks!

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -97,6 +97,79 @@ impl TransactionBuilder {
+        gas_payment: Option<Vec<ObjectID>>,
+        gas_sponsor: Option<SuiAddress>,
+    ) -> Result<TransactionData, anyhow::Error> {
+        let input_gas = if let Some(ref gas_payment) = gas_payment {
+            gas_payment.first()
+        } else {
+            None
+        };
+        let gas_payment = self
+            .select_gas(
+                sender,
+                input_gas.copied(),
+                gas_budget,
+                input_coins,
+                gas_price,
+            )
+            .await?;
```
I don't understand what this is doing, it seems like if you explicitly asked to pay for a transaction using multiple coins, it could potentially ignore the coins you selected after the first one and pick new ones from your wallet? That seems like the wrong thing to do.

stefan-mysten replied:
I am not sure what's unclear here... The overall logic is a bit convoluted, I agree.

The `select_gas` function will use the `input_gas` if it is passed. If it is not passed, then it will get all the owned objects and try to find a gas coin that is not part of the `input_coins`.

amnn replied:
Looking at how this actually gets used, it makes me think that the CLI doesn't actually support gas smashing, and this function is only ever given zero or one gas coin. If that's the case, then the right thing to do is to change the signature so that it doesn't seem like it can take multiple coins only to ignore all but the first one, but if that's not the case, it means the CLI claims to support gas smashing, but doesn't, which is a bug.

stefan-mysten replied:
See https://github.com/MystenLabs/sui/pull/16947#issuecomment-2065049716

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -170,14 +289,23 @@ impl TransactionBuilder {
-        let gas = self
-            .select_gas(signer, gas, gas_budget, input_coins, gas_price)
-            .await?;
+        Ok(coin_refs)
```
nit: these are not really coins anymore, they are arbitrary objects.

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -463,6 +620,73 @@ impl TransactionBuilder {
+                    mutable: true,
+                },
+                Owner::Immutable => {
+                    return Err(anyhow::anyhow!(
```
nit: use `bail`.

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -501,6 +725,55 @@ impl TransactionBuilder {
         )
     }
 
+    /// Construct a transaction kind for the SpliCoin transaction type
```
```suggestion
    /// Construct a transaction kind for the SplitCoin transaction type
```

amnn reviewed (commented):
This is looking like it's on a good track, but I'm not sure about how we're handling coin selection at the moment.

Also, this is a pretty widespread refactoring touching some tricky implementations (`pay-sui`, `transfer-sui` etc) -- what's the testing story?

stefan-mysten commented:
> @stefan-mysten I see that a couple of comments have been marked as resolved, but the accompanying change is not there -- I think you may need to push your changes?

Correct, I did not push yet as I wanted to make sure I added the tests before pushing everything and asking for another set of reviews.

stefan-mysten commented:
> This is looking like it's on a good track, but I'm not sure about how we're handling coin selection at the moment.
> 
> Also, this is a pretty widespread refactoring touching some tricky implementations (`pay-sui`, `transfer-sui` etc) -- what's the testing story?

I added one test for dry runs that also checks if the passed gas object is the one used for paying gas. In addition, I added some assert_eq for those calls that pass a gas object, which assert if the passed `gas obj id` is the same as the one from the `response.effects.gas_object`.

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -180,6 +270,45 @@ impl TransactionBuilder {
+        &self,
+        input_coins: &[ObjectID],
+    ) -> Result<
+        Vec<(
+            ObjectID,
+            sui_types::base_types::SequenceNumber,
+            sui_types::digests::ObjectDigest,
+        )>,
```
nit: You should be able to use the `ObjectRef` type alias here.

stefan-mysten commented:
> Looking at how this actually gets used, it makes me think that the CLI doesn't actually support gas smashing

I don't think the CLI ever supported gas smashing. This `select_gas` function is from the `transaction-builder` crate, which finds the first coin containing at least the `gas_budget` passed. If there's no coin with at least that gas, it will fail. In the `sui-sdk`, it is the same, just tries to find a coin that has enough gas. 

I'll try to fix the signature and then we can add this `gas-smashing` feature on our to-do list.

stefan-mysten commented:
> Imagine I pass multiple coins to gas_payment. input_gas is only the first coin in that list (incidentally this is also not safe because what happens if I pass Some(vec![])?). 
From your description, it sounds like in that case, select_gas will just return back input_coin which is just that first coin, and the remaining coins will be ignored, which does not seem correct.

This is how it used to work before: 
https://github.com/MystenLabs/sui/blob/main/crates/sui-transaction-builder/src/lib.rs#L240
It takes out the first coin from `input_coins`, and in the `new_pay_all_sui` call it adds it to the list at index 0: https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/transaction.rs#L1753

I find this weird, but I tried to keep things as they were, as it is kinda hard to reason about it.

If `gas_payment` is `Some(vec![])`, the `input_gas` variable in `tx_data` function will be `None`, so `select_gas` will iterate through the owned objects to find a coin with that much gas. 

But all in all, the actual semantics of this is simple: it will take the gas object that was passed, or find the first gas coin owned that its value is equal or higher than the passed gas budget.

stefan-mysten commented on crates/sui/src/client_commands.rs:
```diff
@@ -1415,9 +1633,32 @@ impl SuiClientCommands {
+                        vec![],
+                        gas_budget,
+                        gas_price,
+                        input_coins.first().copied(),
```
You are right, it was not working correctly. Thanks for catching this and for the suggestions, now I fixed it! Adding tests and then final review.

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -63,12 +63,12 @@ impl TransactionBuilder {
+        gas_budget: u64,
         input_objects: Vec<ObjectID>,
         gas_price: u64,
     ) -> Result<ObjectRef, anyhow::Error> {
```
As this is a private function, you can probably clean its interface up so that it 

- Takes a `Vec<ObjectID>` of input gas coins and returns a `Vec<ObjectRef>`.
- Accepts a `TransactionKind` instead of a list of input objects, and it does the input object calculation.

stefan-mysten replied:
The issue with this is that all the "original" tx related calls are using it, and those are used by the `json-rpc` so I am a bit scared in changing this. Most of the `TransactionData` functions only take a ObjectRef for gas payment, and not a vec, so not sure what's the improvement there.

stefan-mysten replied:
That would also require to create a `tx_kind` for all those original `tx` functions and pass that into the `select_gas`. Let's do this in a follow-up PR if we really want to.

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -1000,19 +1043,38 @@ impl SuiClientCommands {
+                        tx_kind,
                         gas_budget,
+                        gas_price,
+                        gas.map(|x| vec![x]).unwrap_or_default(),
```
You can do this as follows instead:

```suggestion
                        gas.into_iter().collect(),
```

amnn replied:
(Same goes for all the  examples below).

amnn commented on crates/sui/src/client_commands.rs:
```diff
@@ -1249,18 +1336,51 @@ impl SuiClientCommands {
-                let tx_data = construct_move_call_transaction(
-                    package, &module, &function, type_args, gas, gas_budget, gas_price, args,
-                    context,
+                let tx_kind = construct_move_call_transaction_kind(
```
This function feels a bit redundant now -- it's just wrapping the call to the builder.

amnn reviewed (approved):
Nice! One more small clean-up suggested around `select_gas` but this is now all correct and good to go. Thanks @stefan-mysten!

stefan-mysten commented on crates/sui/tests/cli_tests.rs:
```diff
@@ -3344,13 +3345,23 @@ async fn test_dry_run() -> Result<(), anyhow::Error> {
 
-#[sim_test]
-async fn test_pay() -> Result<(), anyhow::Error> {
+async fn test_cluster_helper() -> (
```
🤦  of course that's cleaner, will do!
